batch/batch/front_end/front_end.py

```diff
@@ -337,15 +357,22 @@ async def _query_batch_jobs_for_billing(request, batch_id):
     return jobs, last_job_id
 
 
```

Comment:
If I understand correctly, keeping '_query_batch_jobs_for_billing' and renaming '_query_batch_jobs' to '_query_job_group_jobs'.
Diff output:

```
diff --git a/batch/batch/front_end/front_end.py b/batch/batch/front_end/front_end.py
index 56a6bcb9b..8a71c4815 100644
--- a/batch/batch/front_end/front_end.py
+++ b/batch/batch/front_end/front_end.py
@@ -268,6 +268,102 @@ async def _handle_api_error(f: Callable[P, Awaitable[T]], *args: P.args, **kwarg
         raise e.http_response()
 
 
+<<<<<<< HEAD
+async def _query_batch_jobs_for_billing(request, batch_id):
+    db = request.app['db']
+
+    # batch has already been validated
+    where_conditions = ['(jobs.batch_id = %s)']
+    where_args = [batch_id]
+
+    last_job_id = request.query.get('last_job_id')
+    query_limit: str = request.query.get('limit')
+    limit = 300
+    if query_limit:
+        try:
+            limit = int(query_limit)
+        except ValueError as e:
+            raise web.HTTPBadRequest(reason=f'Bad value for "limit": {e}')
+    if not (0 < limit < 1e4):
+        raise web.HTTPBadRequest(reason=f'Limit must be between 1 and 10,000 (limit={limit})')
+
+    if last_job_id is not None:
+        last_job_id = int(last_job_id)
+        where_conditions.append('(jobs.job_id > %s)')
+        where_args.append(last_job_id)
+
+    sql = f'''
+    SELECT
+        jobs.batch_id as batch_id,
+        jobs.job_id as job_id,
+        jobs.state as state,
+        batches.user AS user
+    FROM jobs
+    INNER JOIN batches ON jobs.batch_id = batches.id
+    WHERE {' AND '.join(where_conditions)}
+    GROUP BY jobs.batch_id, jobs.job_id
+    ORDER BY jobs.batch_id, jobs.job_id ASC
+    LIMIT %s;
+    '''
+
+    jobs = [dict(record) async for record in db.select_and_fetchall(sql, (*where_args, limit))]
+    n_job_ids = len(jobs)
+    job_ids = [job['job_id'] for job in jobs]
+
+    if n_job_ids == 0:
+        return [], None
+    if n_job_ids == 1:
+        job_condition = 'job_id = %s'
+    else:
+        placeholders = ', '.join(['%s'] * n_job_ids)
+        job_condition = f'job_id IN ({placeholders})'
+
+    job_attributes_sql = f'''
+    SELECT job_id, `key`, `value`
+    FROM job_attributes
+    WHERE batch_id = %s AND {job_condition};
+    '''
+
+    job_resources_cost_sql = f'''
+    SELECT agg.job_id, r.resource, agg.`usage`, r.rate * agg.`usage` as total
+    FROM aggregated_job_resources_v2 agg
+    INNER JOIN resources r ON r.resource_id = agg.resource_id
+    WHERE agg.batch_id = %s AND agg.{job_condition}
+    '''
+
+    attributes_by_job = collections.defaultdict(dict)
+    async for record in db.select_and_fetchall(job_attributes_sql, (batch_id, *job_ids)):
+        attributes_by_job[record['job_id']][record['key']] = record['value']
+
+    resources_by_job = collections.defaultdict(lambda: collections.defaultdict(float))
+    totals_by_job = collections.defaultdict(lambda: collections.defaultdict(float))
+    async for record in db.select_and_fetchall(job_resources_cost_sql, (batch_id, *job_ids)):
+        resource = record['resource']
+        job_id = record['job_id']
+        # Add, in case there are multiple instances of resources with the same name
+        resources_by_job[job_id][resource] += record['usage']
+        totals_by_job[job_id][resource] += record['total']
+
+    for j in jobs:
+        job_id = j['job_id']
+        j['resources'] = resources_by_job.get(job_id, {})
+        # override cost if it already exists
+        j['cost'] = totals_by_job.get(job_id, {})
+        j['attributes'] = attributes_by_job.get(job_id, {})
+
+    last_job_id = None
+    if len(jobs) == limit:
+        last_job_id = jobs[-1]['job_id']
+
+    return jobs, last_job_id
+
+
+async def _query_batch_jobs(
+    request: web.Request, batch_id: int, version: int, q: str, last_job_id: Optional[int]
+||||||| dc7fce0e3
+async def _query_batch_jobs(
+    request: web.Request, batch_id: int, version: int, q: str, last_job_id: Optional[int]
+=======
 async def _query_job_group_jobs(
     request: web.Request,
     batch_id: int,
@@ -276,6 +372,7 @@ async def _query_job_group_jobs(
     q: str,
     last_job_id: Optional[int],
     recursive: bool,
+>>>>>>> 13de4e6ff
 ) -> Tuple[List[JobListEntryV1Alpha], Optional[int]]:
     db: Database = request.app['db']
 
@@ -295,6 +392,84 @@ async def _query_job_group_jobs(
     return (jobs, last_job_id)
```

Reply:
Yep, our local patches add a new function `_query_batch_jobs_for_billing()` and there's been a conflict because they've renamed the adjacent function. So we need to keep their renaming and keep our added function.

However there's a bit more we should do too: the change to the adjacent function (and the rest of the file) is to refactor due to the big internal change from batches to job groups. So we may need to look at whether our new function `_query_batch_jobs_for_billing()` needs some adjusting for job groups too. Which will not be fun…

Reply:
So if the original table aggregated_job_resources_v2 is still being populated, then we would not need to modify anything I assume. I can see aggregated_job_resources_v2 is used side by side with aggregated_job_resources_v3 in the previous versions of hail.